getAllCourseData.py

```diff
@@ -0,0 +1,26 @@
+from flask import Flask, request
+import requests
+import json
+app = Flask(__name__)
+@app.route('/',methods = ['GET'])
+URL = 'http://api.culpa.info/courses/department_id/7'
+def getNames():
+    r = requests.get(URL)
+    jason_data = json.loads(r.text)
+    array = ""
+    for i in range(len(jason_data['courses'])):
+        array = array + str(jason_data['courses'][i]['name'])+'x'
```

Comment:
We should really replace this with:

```python
array = "x".join(course["name"] for course in jason_data["courses"])   # Do we need the "x" at the end?
```
As is, we're doing an `O(n^2)` quadratic . Everytime you run `+`, Python has to allocate a new string, which means going through an expensive `malloc` and `free` call (not to mention the reference counting overhead) on each intermediary string, which adds up really quickly. Calling `join` allows Python to allocate only once up front, which is _much_ faster.

So, as a concrete example, suppose each `name` is 10 bytes long, and there are 80 `names` we're concatenating. On the first loop iteration, the original loop will allocate `10 + 1 = 11` bytes for `array`, the second will allocate `2 * 11 = 22` bytes, the third will allocate take `3 * 11 = 33` and so on. Adding up `11 + 22 + 33 + ... + 80 * 11`, we allocate more than `35640` bytes, even though the final string only takes `880` bytes!

But by calling `join`, Python will do some clever optimizations and only allocate the final `880` bytes instead. Much faster.

For more reading on this:
- http://www.joelonsoftware.com/articles/fog0000000319.html
- https://ericlippert.com/2013/06/24/string-concatenation-behind-the-scenes-part-two/